### PR TITLE
Added archived details while archiving user

### PR DESCRIPTION
**Pull request by gauravsinhaweb**
### This PR is linked to Issue no. #1214: Added archived user details during the archiving process

**API contract link: https://github.com/Real-Dev-Squad/website-api-contracts/pull/138**
**Data Modal Link: https://github.com/Real-Dev-Squad/website-data-models/pull/48**

**Changes Made:**
- added a field in db `archivedDetails` if the user is archived. 
- added test cases
### Structure: 
```js 
{
  archivedDetails: [{
    updated_at: Number,
    super_user_id: String,
    reason: String
  }],
  username: String,
  github_display_name: String,
  github_id: String,
  isMember: Boolean,
  roles: {
    member: Boolean,
    in_discord: Boolean,
  },
  tokens: {
    githubAccessToken: String,
  },
  incompleteUserDetails: Boolean,
}
```

The reason needs to pass in the body as JSON object in the PATCH request:

**Example: **

```json
{
    "reason":"some reason"
}
```

✅ Test Case Results:

Screenshot 2023-07-10 085443

Screenshot 2023-07-10 085507

### CODE REVIEW

**Review by heyrandhir — CHANGES_REQUESTED** [review]
please write the tests. Or get a approval from Ankush to get it merged without tests 👀

**Review by bhtibrewal — CHANGES_REQUESTED** [review]
Also please add the updated API contract and model with this PR

**Comment by gauravsinhaweb** [comment]
> Also please add the updated API contract and model with this PR

API contract PR : https://github.com/Real-Dev-Squad/website-api-contracts/pull/138

**Comment by bhtibrewal** [comment]
> > Also please add the updated API contract and model with this PR
> 
> API contract PR : [Real-Dev-Squad/website-api-contracts#138](https://github.com/Real-Dev-Squad/website-api-contracts/pull/138)

please get review on your API contract and the data model needs to be updated in https://github.com/Real-Dev-Squad/website-data-models

**Comment by gauravsinhaweb** [comment]
hey, @heyrandhir @bhtibrewal I've added test cases, data modelling and API contacts, please review it, thanks

File: controllers/members.js [hunk 1]
```diff
@@ -85,8 +86,15 @@ const archiveMembers = async (req, res) => {
   try {
     const { username } = req.params;
     const user = await fetchUser({ username });
+
+    const userId = user.user.id;
+    const currentUserID = req.userData.id;
+    const body = req.body;
+    if (body?.reason === "" || body?.reason === undefined) {
```
Comment (iamitprakash) [inline controllers/members.js:93]: reason can be " " as well

File: models/members.js [hunk 2]
```diff
@@ -99,15 +99,25 @@ const fetchUsersWithRole = async (role) => {
...
+const addArchiveRoleToMembers = async (userId, currentUserId, reason) => {
   try {
     const userDoc = await userModel.doc(userId).get();
     const user = userDoc.data();
     if (user?.roles && user.roles[ROLES.ARCHIVED]) return { isArchived: true };
     const roles = { ...user.roles, [ROLES.ARCHIVED]: true };
-    await userModel.doc(userId).update({
-      roles,
-    });
+    const archivedDetails = {
```
Comment (iamitprakash) [inline models/members.js:108]: here where you set the archivedDetails property. The value of the reason property is being set to the empty string if the user does not provide a reason for archiving the user. However, the reason property is a required property, so the empty string is not a valid value.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: controllers/members.js (hunk 1 above)
Comment (gauravsinhaweb) [reply controllers/members.js]: updated!, use regex to verify empty/ undefined reason

File: models/members.js (hunk 2 above)
Comment (gauravsinhaweb) [reply models/members.js:108]: if the reason property is an empty string, it will throw an error of `400` bad request, reason required

**Pull request closed without merging**